lib/internal/streams/from.js

```diff
@@ -34,21 +34,52 @@ function from(Readable, iterable, opts) {
     objectMode: true,
     ...opts
   });
+
   // Reading boolean to protect against _read
   // being called before last iteration completion.
   let reading = false;
+
+  // needToClose boolean if iterator needs to be explicitly closed
+  let needToClose = false;
+
   readable._read = function() {
     if (!reading) {
       reading = true;
       next();
     }
   };
+
+  readable._destroy = async function(error, cb) {
+    try {
+      if (needToClose) {
+        needToClose = false;
+        await close();
+      }
+    } catch (e) {
+      error = error || e;
+    } finally {
+      process.nextTick(cb, error);
+    }
+  };
```

Comment:
Restoring my objection on having this as an async function.

Reply:
Not opposed. But curious as to why?

Reply:
The signature of this function does not return a promise. The caller of _destroy() is not expecting a promise to be returned, while this implementation is ok, it encourages a pattern that should be avoided.

Reply:
Sorry to not clearly understand your feedback. I really appreciate your patience.
Pushed.